### ISSUE TITLE
[react/no-unknown-property] rule conflicts with `@react-three/fiber` on `<mesh />` and other tags

### ISSUE THREAD

**Issue by JustFly1984** [jsx-eslint/eslint-plugin-react#3423]
I'm using `@react-three/fiber` in my app, and updating `eslint-plugin-react` broke the linting.

You did not took into account custom renderers while introducing this changes.

on `<mesh />` false negatives:

Unknown property 'castShadow' found
Unknown property 'receiveShadow' found
Unknown property 'dispose' found
Unknown property 'geometry' found
Unknown property 'userData' found
Unknown property 'rotation' found

on `<meshStandardMaterial />` false negatives:

Unknown property 'metalness' found
Unknown property 'roughness' found
Unknown property 'side' found

on `<planeBufferGeometry />`

Unknown property 'args' found

on `<meshPhongMaterial />`

Unknown property 'flatShading' found
Unknown property 'shininess' found
Unknown property 'side' found
Unknown property 'specular' found

and a lot more.

Please fix support for `@react-three/fiber`

Meanwhile I had to disable `react/no-unknown-property` rule

**Comment by ljharb** [comment]
Since `<mesh>` is not a standard tag, and isn't a custom element, it shouldn't be used at all by that library.

That said, since React ignores nonstandard tags, so should we, which makes this a duplicate of #3420.

**Comment by JustFly1984** [comment]
I also agree that it is a non standard tag, but three.js has massive user base, so could it be a separate option to support react-three-fiber tags and attributes? They do have good typescript support.

**Comment by ljharb** [comment]
I don't think it's worth the complexity - there's multiple libraries that do things they shouldn't by introducing invalid tags. Three.js may want to build an eslint plugin just to check their own tags, though - then they can own their own decisions.

**Comment by vincent-lecrubier-skydio** [comment]
> there's multiple libraries that do things they shouldn't

Who says they shouldn't? Many Custom react renderers exist and I am not aware of anyone saying this shouldn't be done. If anything, the react team has been supportive.

**Comment by iamtofr** [comment]
i think u should define these props in a rule to allow them like: 

"react/no-unknown-property": [1, { "ignore": ["position"] }]

**Comment by brianzinn** [comment]
Hi, Is this going to be accepted as an issue?  I'm trying to work out how to support in a renderer for another 3d engine.
https://github.com/brianzinn/react-babylonjs/issues/249

I've added to the JSX Intrinsic Elements from my library and VS code has intellisense and works very well with these custom properties.  There are a lot of renderers that are NOT rendering HTML, so they will also start to fail...

**Comment by ljharb** [comment]
@brianzinn there's nothing to accept - the rule already has an `ignore` config, which you can use to ignore any nonstandard properties you like.

If you want free reign to name properties anything you like, then components need to be named in PascalCase, since lowercased components in react are only HTML elements.

**Comment by brianzinn** [comment]
ok - thanks for responding - appreciate that.  I'll probably need to ignore 1,000 properties.  Unfortunately it looks like you don't understand how renderers work. PascalCase are components handled by react.  lowercase components are intrinsic elements handled by the renderer (not just the DOM renderer). In other words, lowercased components in react are *NOT* only HTML elements.  Your profile says you are obsessed with backwards compatibility, so bit surprising, but thanks again for clarifying!  cheers.

**Comment by ljharb** [comment]
I understand how renderers work; we only support the react renderer in eslint-plugin-react.

**Comment by coobeeyon** [comment]
@ljharb would you mind ELI5 what is the difference between a `<mesh>` and any other React component which I author?  I seem to have no problem making custom props for my own stuff.  Why wouldn't `<mesh>` be able to do so too?

**Comment by CodyJasonBennett** [comment]
> I understand how renderers work; we only support the react renderer in eslint-plugin-react.

This would more specifically be strictly `react-dom`, and this is not future proof to custom elements support. For those working in an x-platform codebase, TypeScript would be of more help here.

**Comment by ljharb** [comment]
@coobeeyon custom compoments are defined by their author; html elements are defined by the browser.

**Comment by CodyJasonBennett** [comment]
Not cool. This is needlessly contrarian and frankly unprofessional. If you only want to support a subset of JSX that's fine, but don't play this blame game with anything outside of this model.

**Comment by ljharb** [comment]
@CodyJasonBennett I’m not sure what you mean. Anything lowercased either conforms to the html standard or is invalid, incorrect, and bad - that’s just the way react works.

**Comment by CodyJasonBennett** [comment]
That is not true. I suggest learning how React and host components work; React is x-platform in nature.

**Comment by ljharb** [comment]
I’m aware that react 19 may support custom html elements, but I’m not aware of react supporting that yet.

Either way, web components in jsx simply wouldn't be statically analyzable, so if you’re using those, you’re probably not going to get much benefit from eslint.

**Comment by coobeeyon** [comment]
> 

Well the components in question are certainly designed by the author not the consortium. Shouldn't that mean they can define their own properties?  Is the issue really that they are lower case?

**Comment by ljharb** [comment]
The grammar of jsx says that lowercased elements are *solely* the province of the renderer, and it is simply not possible for authors to define them, so yes, that is the entire issue. Components created by developers must ALWAYS start with an uppercase letter; components starting with a lowercase letter belong *entirely* to the renderer.

**Comment by kaloyanBozhkov** [comment]
+1 on this, I am on the same boat

**Comment by rivertam** [comment]
React isn't a renderer but a reconciliation algorithm. react-three-fiber isn't doing something terribly niche with React; it's just not react-dom. React Native and projects like react-blessed have been using this renderer-agnostic aspect of JSX/React for far longer than React 19; I'm surprised this didn't come up in November. It's certainly not true that anything lowercase conforms to the html standard; I don't even think that's true for react-dom.

I don't really know how this relates to eslint in particular, but yeah this basically makes this plugin `eslint-plugin-react-dom`, not `eslint-plugin-react`.

**Comment by ljharb** [comment]
In react, anything lowercase can’t be a custom component, which is why React Native uses non-lowercase components for its primitives like Text etc.

React *is* react-dom (RN specifically doesn’t clash with react-dom, on purpose). It’s not eslint-plugin-jsx, it’s eslint-plugin-react, and that means the react ecosystem. This isn’t a renderer-agnostic plugin, nor are most things in the ecosystem.

You are welcome to disable the rule if it doesn’t work for you.

**Comment by JustFly1984** [comment]
The thing is that it works great in all other parts of an app, except custom renderers. It is annoying to add eslint ignore comments for every non html intrinsic element.

**Comment by ljharb** [comment]
I can certainly understand the annoyance - the "working ok for custom renderers", unfortunately, was "not working properly for the standard renderer".

Do you have a better suggestion beyond listing elements in `ignore` that wouldn't be a footgun for the primary use case?

**Comment by JustFly1984** [comment]
is it possible to ignore non-standard lowercase primitives? is it possible to extend the list of primitives, providing valid attributes list for each of them?

**Comment by ljharb** [comment]
The former would be a potential foot gun - the latter, though, seems workable, albeit complex.

**Comment by JustFly1984** [comment]
I guess each non-standard renderer eventually will provide their own config to extend the rule, if there is an API to satisfy the need.

**Comment by ljharb** [comment]
Unfortunately eslint rule configuration doesn’t compose, so you’d have to manually merge together various configs :-/

**Comment by JustFly1984** [comment]
this is advanced stuff, so I guess advanced people can handle to manually setup the config by copy-pasting somebody's solution.

**Comment by drcmda** [comment]
i didn't catch this thread but this assumption

> I’m aware that react 19 may support custom html elements

is 100% wrong @ljharb. lowercase in react has nothing, whatsoever to do with html or custom html elements. it designates a native host element. the renderer defines these for its host platform. react-dom defines divs and spans, other renderers define theirs. a `<mesh>` is a new THREE.Mesh(), it doesn't need a browser or html to function.

this is fundamentally how react works, not react 19 but just react.

**Comment by ljharb** [comment]
@drcmda yes, I’m aware of that, and the only renderer this rule supports is react-dom. Otherwise, how could we have “unknown” or known properties on an unknown renderer?

**Comment by cyrilchapon** [comment]
Just landed here, coming from [here](https://github.com/pmndrs/react-three-fiber/discussions/2487), with the exact same use-case.

Just a developing-user point of view :
- I have a setup which involves both dom and fiber renderers, at the same time
- I'm OK to go for advanced and complex eslint configuration for this use-case (anyone using eslint and eslint-dom with any-transpiler + any-bundler + any-opinion already have a 120 lines eslintrc.json; this argument is a nonsense)
- I'd expect react-three-fiber to provide some hints on how to configure this
- This would apparently require some tweaking into that rule

> is it possible to extend the list of primitives, providing valid attributes list for each of primitive? preferably specifying attribute value types
This seems like the natural way to go. That would unlock react-three-fiber / [any other renderer team] to provide such a list to go in there.

> React is react-dom (RN specifically doesn’t clash with react-dom, on purpose)
Well you are basically just wrong, and well it happens that another package, react-three-fiber, does clash with react-dom. On purpose. It's written as a renderer should be; the `react/no-unknown-property` rule is not written as it should be.

When the community needs here is a fix, not a random opinion about renderer casing.

> The former would be a potential footgun - the latter, though, seems workable, albeit complex.
I can give it a shot, though I never contributed here. What do you mean "complex" ? Any chance you can point the right direction to start ?

Thanks

**Comment by ljharb** [comment]
The react/no-unknown property makes the perfectly reasonable and virtually universally true assumption that react-dom is being used for non-custom JSX components.

The complexity is in designing a schema that can cover "any random hallucination some renderer can come up with" while still being trivial to configure to common use cases, and while avoiding confusing, slow, and hard to maintain code in the rule itself.

**Comment by cyrilchapon** [comment]
Baby steps are the key here 🙂 9 months after the initial issue, you're beginning to consider the root of the trouble might after-all be just an "assumption" and not that "truth from the standards god" you argued in this thread about 10 times. 

That's a great start !

Your nonsense-adverbs skills in sentences are quite impressive; but in the real world, that assumption — even if indeed "reasonable" and "true" (please notice how it's much less impressive without the nonsense adverbs) is **completely off-topic**.

-------

Let's try another approach, maybe to speed up the intellectual process just a little bit — after all, we can admit actual progress after those 9 months :

## => ✨ `@react-three/fiber` **is not rendering to react-dom** !!!  ✨

-------

> The complexity is in designing a schema that can cover "any random hallucination some renderer can come up with" while still being trivial to configure to common use cases, and while avoiding confusing, slow, and hard to maintain code in the rule itself.

Thanks for this quite constructive "starting point" I asked; for sure a monkey couldn't have told the same answer. Great progress here too !

In the real world though :

```json
{
  "tagsMapping": {
    "@react-three/fiber": ["mesh", "..."]
  }
}
```

**Comment by ljharb** [comment]
@cyrilchapon I've never argued anything like "truth from the standards god", and frankly your patronizing tone is hostile and unwelcome.

Your tone makes me want to state that this rule will NEVER, EVER accommodate this use case. Is that the outcome you're hoping for?

**Comment by cyrilchapon** [comment]
@ljharb at least this gets some attention.

That patronizing tone might just be an answer to the childish tone... ?

Come on man, I'm basically repeating an argument that has been stated here like 20 times. 

Be assured that was nothing more than trying to make it funny though, as a desperate call.

**Comment by ljharb** [comment]
Not the kind of attention you want, I assure you.

**Comment by cyrilchapon** [comment]
Well grow up, or at least rename that joke eslint-react-dom-plugin.

**Comment by FalconPilot** [comment]
I can predict the appearance of a new package, probably called `jsx-eslint-fiber`, just because this one doesn't support one rule... :/

I can understand the reasoning behind the complexity of implementation, but when the reasoning doesn't correspond to the real-world application, isn't it a problem on the long run?

**Comment by JustFly1984** [comment]
Guys, I'm a topic starter, I'm also a maintainer of open source project, I do it more than 5 years. I've burned out not once, and there is not much people who has time or skills to contribute. There is also not much donations, or other ways to monetize.

This is an open source, you can fork this package, change the rule, and publish it under different name, and post the link here. But be mindful that you will have to maintain it in long run.

I'm not using react-three-fiber in my current projects so I have no incentive. But for those who use it, this issue is just a huge bummer, especially for juniors. 3d is hard enough, even without tooling issues.

**Comment by ljharb** [comment]
I agree - and it's unfortunate for those users that fiber would choose to use lowercase tag names, instead of capitalizing what are custom component names.

**Comment by brianzinn** [comment]
I don't see the casing as unfortunate.  That's how host elements are differentiated between React Components.  It's unfortunate that unknown `elements` aren't just ignored automatically - from my experience it's confusing that this plugin complains about some properties, so it's not a good DX.  Likely most react renderers include their own JSX Intrinsic Elements, so it's not even strictly necessary for eslint as editors will pick up with intellisense/autocomplete?

**Comment by ljharb** [comment]
Hmm, that actually might be a reasonable compromise. Is there any overlap between fiber's made-up element tag names and HTML element tag names? If not, we could add an option to `ignoreUnknownHostElements`.

**Comment by brianzinn** [comment]
Thank-you @ljharb for being open to more dialog.  In the case of my react renderer - I renamed any known ones to avoid overlap.  So, I don't use known element names like `button`, `image`, `line`, etc - quite a few conflicted ie: `polygon` with `React.SVGProps<SVGPolygonElement>`.

I think the option is a really good idea, but would prefer an opt-out mechanism as opposed to opt-in (ie: default enabled).  With the advent of Web Components and other custom elements feel that the linting is best used for known HTML elements such as div, span, etc.  I can't speak for other renderers, but if they overlap with known elements then that's for them to fix I think...

**Comment by ljharb** [comment]
@brianzinn to avoid a breaking change it would have to be opt-in. I generally agree that if there's an overlap with html it's a bug in the custom renderer, but it's still useful to know if there's any existing overlaps to worry about.

@JustFly1984 It's not yet clear that's the right path, but if we decide on a direction then i'll do so.

**Comment by FalconPilot** [comment]
I think that, reading the issue's history, one of the problems has been to mix up HTML W3C standards with JSX/React/React-dom standards.

In theory, that's a sound strategy, as in convention, core HTML nodes are lowercased, and custom JSX components are CamelCased. However, we must keep in mind that this is a *convention*, not an actually strict rule if you use regular JS. *However*, things go even further when we start using TypeScript...

People using TypeScript will be familiar with the following:

<img width="550" alt="Capture d’écran 2023-05-02 à 22 26 29" src="https://user-images.githubusercontent.com/9877634/235778218-9d1c4843-5b8a-4cab-bdee-78fd25457904.png">

In that example, I created a simple `blah` FunctionComponent, and rendered it into a `Foo` component. What happens, then? The above error, in TypeScript.
```tsx
const blah: React.FC = () => {
  return (
    <div>Blah</div>
  )
}

export const Foo: React.FC = () => {
  return (
    <blah/>
  )
}
```
My hot take on it would be that JSX actually considers ***any lowercased component to only be contained within the JSX core library*** and that any library willing to implement lowercased-components have to override this definition. I'm merely speculating, but my point is, this is another case where conflicting views on how the ecosystem *should* work is the root problem here.

In a world where the tooling is most of the time adding more complexity due to the inherent nature of the projects it's bound to, I think that just either removing/opting-in problematic rules *or* simply letting the users override them themselves are the two easiest, but also most realistic solutions.

... Or we could all start programming our web apps in Elm, PureScript or ReasonML and not have to deal with this anymore. Hey, I have a right to dream, right? ¯\_(ツ)_/¯

**Comment by brianzinn** [comment]
@ljharb - Is it not a breaking change that unknown elements are currently errors - isn't that what created this issue in the first place?  I don't recall having this issue until recently, so it seems like opt-out would be backwards compatible with the original behaviour - please correct me if I'm wrong there!! 😄 

@FalconPilot My understanding is that the JSX.IntrinsicElements are typically extended. ie:
```typescript
declare global {
  namespace JSX {
    interface IntrinsicElements {
      blah: YourTypeHere
   }
  }
}
```

**Comment by FalconPilot** [comment]
@brianzinn Correct, however, it needs to be explicitly extended, which adds another layer of complexity 🤔 Because it means that the library implementing custom, lowercased components has to extend it *and* the Linter has to acknowledge those new components, which is, IMHO, way too hazardous.

The "CamelCase" convention, while handy, showed its limits as soon as not everyone seemed to adopt it, alas - that's the problem of a convention being only implicitly shared among a developers community. But hey, that problem is pretty much JS in a nutshell, so I'm not pointing fingers at anyone specifically

**Comment by ljharb** [comment]
@brianzinn in the sense that any bugfix is a breaking change, sure - but this was how the rule was always meant to behave.

@FalconPilot Regarding jsx, it's not a convention - the jsx specification says that lowercase tag names are "host" elements, which means they indeed delegate to the renderer.

The _react_ renderer (including RN), however, has made host elements largely be HTML (and sometimes SVG) elements. This means that when a custom renderer chooses a different semantic for host elements (which jsx allows!) then it's no longer compatible with react-specific jsx - which is what this plugin is solely focused on.

**Comment by FalconPilot** [comment]
@ljharb I would agree with you, save for the fact that it isn't enforced by any mean 🤔 Which means that it is only a "recommendation", not a strict rule.

If JSX crashed at runtime if a custom component was declared using a lowercased name, that would be extreme, but at least it would put everyone on the same page. *Or*, if the problem lies in the *React* implementation, then it could be React's JSX implementation that could crash it. My point is, "if nothing forces the specification to be respected, then someone is bound not to respect it and stir up trouble".

A specification that can simply be ignored is, to me, just kinda mystical. Well, it can be ignored in JS, at least, in TS, it's enforced via the compiler.

The (naively) ideal fix, IMHO, would be on `fiber.js`'s side:
1. Create CamelCased components in lieu of lowercased ones
2. Keep the lowercased ones, but only as aliases for the new, CamelCased ones
3. Add a console warning to warn about depreciation of lowercase components
4. Profit

I'm not using it currently as I'm not working in any WebGL 3D related project, but if some people want to suggest it to them, maybe that'd be a good idea?

**Comment by ljharb** [comment]
It's enforced by react itself, which uses react-dom - it is thus a strict rule _for react_. If you're using a custom renderer you are risking that react tools won't work with your source code.

I completely agree that the best fix would be for fiber to follow React Native's path and use PascalCased (not lowercased or camelCased) component names for its primitives.

**Comment by FalconPilot** [comment]
Ah, now that I look at it more carefully, the root issue stems from the fact that `fiber.js` actually uses host elements from 3D engines, which technically makes those lowercased perfectly valid 🤔

An opt-in for the ESLint ruleset would therefore perhaps be more reasonable. If the library is bound to a custom renderer that actually is compatible with react itself, then that pretty much settles it.

**Comment by ljharb** [comment]
"valid" for jsx, sure - but not valid for react and react-dom.

**Comment by FalconPilot** [comment]
I can see why for react-dom, but isn't react technically a separate library from react-dom? (unless I missed a beat somewhere)

AFAIK, React is agnostic of its renderer which has to be called explicitly in order to render React nodes 🤔

**Comment by ljharb** [comment]
This plugin is for react with react-dom.

**Comment by FalconPilot** [comment]
Then, wouldn't it be clearer for people if it was called `jsx-eslint-react-dom`? (Although it's too late to rename it now)

**Comment by ljharb** [comment]
Probably it would, but the plugin was created before any alternative renderers even existed, and react will always be predominantly used with react-dom.

Listing something in the readme to the effect of "the primary purpose of this plugin is to support react used with react-dom and React Native; accommodations may be made for custom jsx renderers" would be fine tho.

**Comment by brianzinn** [comment]
I don't think that is accurate this plugin was created before alternative renderers existed.  react-art renderer started before this one.  yannickr started this in 2014 and I see react-art published 10 years ago: https://www.npmjs.com/package/react-art/v/0.0.1?activeTab=versions

I don't personally believe the original author of this library intended it to be for DOM only.  I do agree that react-dom is predominant use-case, but it leaves many especially new users confused and needing to configure out of these changes that aren't backwards compatible.

**Comment by ljharb** [comment]
React itself wasn't widely used until like 2014-2015, but fair enough.

**Comment by JustFly1984** [comment]
And Three.js exists since April 24, 2010

**Comment by ljharb** [comment]
@JustFly1984 oh sure, but not used with jsx :-)

**Comment by JustFly1984** [comment]
JSX was initially released in 2013 as part of the React JavaScript library. I'm glad we don't talk about jQuery.

**Comment by JustFly1984** [comment]
I've tried to use Three.js over the years in different combinations with different frameworks and vanilla JS. 
I can state one simple thing: Until recently Mobile tech was not capable to run 3d in canvas without lag or significant battery drainage. There was no real point to use react + canvas 3d, cos even lowpoly models were very laggy even in perf optimized React and React Native. 
Now with three.js and react you can build whole app in canvas, or 3d world, using custom renderer based on standards evolved from long before all hype around react.js happened. 

You are trying to impose local React.js community standards on global three.js community. Please be attentive that you are guarding the bridge which could connect 2 big communities which had little intersection before due to compute limitations. Now they could unite. please don't keep it harder than it is already. Tooling should not prevent people from using tech in correct way.

I would not be so active on this topic, if this issue has been existing from beginning of times, due to intrinsic incompatibility. But it was introduced 7 months ago in patch version and broke our CI/CD pipeline out of the blue.

**Comment by ljharb** [comment]
I certainly apologize for it breaking your CI; that's never ideal, but this is how the rule was always *meant* to behave.

I very much doubt that one eslint plugin is the "bridge" between three.js and React - someone can always make `eslint-plugin-react-threejs` or something that clones this rule to work with it.

**Comment by brianzinn** [comment]
I don't think this issue is about making a new plugin.  it's about a plugin that is part of the ecosystem of React, but doesn't play nice with other renderers.  We can't, for example, control that it's part of CRA via [eslint-config-react-app](https://github.com/facebook/create-react-app/tree/main/packages/eslint-config-react-app)!  I feel like this conversation isn't going anywhere anyway...

**Comment by ljharb** [comment]
CRA is deprecated anyways and should never have been used for anything but rapid prototyping, or learning.

**Comment by brianzinn** [comment]
sure, but `create-next-app` does...  Anyway, I'm fine to leave it unchanged as surely you are not convinced otherwise 😄 looks like vite@latest does not use this plugin, so that's another "workaround"!

**Comment by JustFly1984** [comment]
This library is definitely a part of eslint-config-react-app, which is propagated to other libraries and projects including nextjs and I think gatsbyjs too. So basicaly this issue would be facing everybody who tries to use custom renderers in any of major 
 frameworks.
I'm sure this issue already killed a ton of time of everybody following it and will kill much more in the future and this thread will continue until somebody do something about it.

**Comment by terry-codecov** [comment]
For anyone running into this issue in the future, the react fiber maintainers have released their own [eslint plugin which fixes the issue.](https://github.com/pmndrs/react-three-fiber/tree/master/packages/eslint-plugin)



(Also thanks to the maintainers of the react eslint plugin, it's been a staple install for me for years!)

**Comment by ljharb** [comment]
@terry-codecov that's awesome! I'm not sure how it fixes the issue; it has 2 rules in it and it's not an alternative parser, but that is absolutely where this kind of issue should be addressed :-)

**Comment by alessiocancian** [comment]
> For anyone running into this issue in the future, the react fiber maintainers have released their own [eslint plugin which fixes the issue.](https://github.com/pmndrs/react-three-fiber/tree/master/packages/eslint-plugin)
> 
> (Also thanks to the maintainers of the react eslint plugin, it's been a staple install for me for years!)

That doesn't fix the issue, it can't replace eslint-plugin-react.

**Comment by cruhl** [comment]
For anyone else trying to work around this rule incorrectly assuming HTML/DOM is the only valid react renderer, here's a yarn patch I've used to get around the issue...

```
diff --git a/lib/rules/no-unknown-property.js b/lib/rules/no-unknown-property.js
index fc573366956db27eca1f9046d465f1f445fa1c2b..9dbb47810b270ef50210d4487c8143d0a7e52b19 100644
--- a/lib/rules/no-unknown-property.js
+++ b/lib/rules/no-unknown-property.js
@@ -10,6 +10,146 @@ const docsUrl = require('../util/docsUrl');
 const testReactVersion = require('../util/version').testReactVersion;
 const report = require('../util/report');
 
+const reactThreeElements = [
+  "object3D",
+  "audioListener",
+  "positionalAudio",
+  "mesh",
+  "instancedMesh",
+  "scene",
+  "sprite",
+  "lOD",
+  "skinnedMesh",
+  "skeleton",
+  "bone",
+  "lineSegments",
+  "lineLoop",
+  "points",
+  "group",
+  "camera",
+  "perspectiveCamera",
+  "orthographicCamera",
+  "cubeCamera",
+  "arrayCamera",
+  "instancedBufferGeometry",
+  "bufferGeometry",
+  "boxBufferGeometry",
+  "circleBufferGeometry",
+  "coneBufferGeometry",
+  "cylinderBufferGeometry",
+  "dodecahedronBufferGeometry",
+  "extrudeBufferGeometry",
+  "icosahedronBufferGeometry",
+  "latheBufferGeometry",
+  "octahedronBufferGeometry",
+  "planeBufferGeometry",
+  "polyhedronBufferGeometry",
+  "ringBufferGeometry",
+  "shapeBufferGeometry",
+  "sphereBufferGeometry",
+  "tetrahedronBufferGeometry",
+  "torusBufferGeometry",
+  "torusKnotBufferGeometry",
+  "tubeBufferGeometry",
+  "wireframeGeometry",
+  "tetrahedronGeometry",
+  "octahedronGeometry",
+  "icosahedronGeometry",
+  "dodecahedronGeometry",
+  "polyhedronGeometry",
+  "tubeGeometry",
+  "torusKnotGeometry",
+  "torusGeometry",
+  "sphereGeometry",
+  "ringGeometry",
+  "planeGeometry",
+  "latheGeometry",
+  "shapeGeometry",
+  "extrudeGeometry",
+  "edgesGeometry",
+  "coneGeometry",
+  "cylinderGeometry",
+  "circleGeometry",
+  "boxGeometry",
+  "capsuleGeometry",
+  "material",
+  "shadowMaterial",
+  "spriteMaterial",
+  "rawShaderMaterial",
+  "shaderMaterial",
+  "pointsMaterial",
+  "meshPhysicalMaterial",
+  "meshStandardMaterial",
+  "meshPhongMaterial",
+  "meshToonMaterial",
+  "meshNormalMaterial",
+  "meshLambertMaterial",
+  "meshDepthMaterial",
+  "meshDistanceMaterial",
+  "meshBasicMaterial",
+  "meshMatcapMaterial",
+  "lineDashedMaterial",
+  "lineBasicMaterial",
+  "primitive",
+  "light",
+  "spotLightShadow",
+  "spotLight",
+  "pointLight",
+  "rectAreaLight",
+  "hemisphereLight",
+  "directionalLightShadow",
+  "directionalLight",
+  "ambientLight",
+  "lightShadow",
+  "ambientLightProbe",
+  "hemisphereLightProbe",
+  "lightProbe",
+  "spotLightHelper",
+  "skeletonHelper",
+  "pointLightHelper",
+  "hemisphereLightHelper",
+  "gridHelper",
+  "polarGridHelper",
+  "directionalLightHelper",
+  "cameraHelper",
+  "boxHelper",
+  "box3Helper",
+  "planeHelper",
+  "arrowHelper",
+  "axesHelper",
+  "texture",
+  "videoTexture",
+  "dataTexture",
+  "dataTexture3D",
+  "compressedTexture",
+  "cubeTexture",
+  "canvasTexture",
+  "depthTexture",
+  "raycaster",
+  "vector2",
+  "vector3",
+  "vector4",
+  "euler",
+  "matrix3",
+  "matrix4",
+  "quaternion",
+  "bufferAttribute",
+  "float16BufferAttribute",
+  "float32BufferAttribute",
+  "float64BufferAttribute",
+  "int8BufferAttribute",
+  "int16BufferAttribute",
+  "int32BufferAttribute",
+  "uint8BufferAttribute",
+  "uint16BufferAttribute",
+  "uint32BufferAttribute",
+  "instancedBufferAttribute",
+  "color",
+  "fog",
+  "fogExp2",
+  "shape",
+];
+
 // ------------------------------------------------------------------------------
 // Constants
 // ------------------------------------------------------------------------------
@@ -546,6 +686,8 @@ module.exports = {
 
         const tagName = getTagName(node);
 
+        if (reactThreeElements.includes(tagName)) { return; }
+
         if (tagName === 'fbt' || tagName === 'fbs') { return; } // fbt/fbs nodes are bonkers, let's not go there
 
         if (!isValidHTMLTagInJSX(node)) { return; }
```

**Comment by cruhl** [comment]
By the way, here is a list of non-DOM renderers this rule simply ignores...
 
 https://github.com/chentsulin/awesome-react-renderer

**Comment by ljharb** [comment]
@cruhl it's not incorrect, it's the purpose of the entire plugin. See https://github.com/jsx-eslint/eslint-plugin-react/issues/3423#issuecomment-1532160390

**Comment by cruhl** [comment]
I think react and it's arbitrary renderers are being conflated:

![CleanShot 2024-02-06 at 17 18 39](https://github.com/jsx-eslint/eslint-plugin-react/assets/4983357/8cd0cda7-7139-4de9-85a5-a65990c983ae)

React does not contain a renderer, ReactDOM does, so for this to be considered the "React plugin" and not the "React DOM plugin" it would need to support arbitrary renderers right?

**Comment by ljharb** [comment]
@cruhl it doesn't necessarily work with React Native either, so yes, it's a React DOM plugin, named "react" because at the time it was created that was the only renderer.

It could potentially handle additional renderers, but that would have to be approached with the attitude of "please add this complexity" and not "i'm entitled to your free labor and how dare you not have already thought of my use case in advance".

**Comment by cruhl** [comment]
I might take a crack at extending my patch to enable arbitrary element/property maps which could be provided as a config.

Would you be interested in adding something like that if I submitted a PR?

I'm thinking of just passing through a custom map like how the defaults are defined:
![CleanShot 2024-02-06 at 17 27 42](https://github.com/jsx-eslint/eslint-plugin-react/assets/4983357/844aed64-fc44-4d46-9891-392df9a57bd8)

**Comment by ljharb** [comment]
Would that custom map replace, or augment, the defaults?

**Comment by cruhl** [comment]
You probably know more about what people would prefer in general.

For my use-case it's just extending the existing set.

Maybe others would want a full replacement?

Seems easy to implement either way once the contract is defined. 👍

**Comment by ljharb** [comment]
Fair.

One last question before I welcome a PR - can you provide an example map that, for example, would solve this issue? (and another one for your use case, if different)

**Comment by cruhl** [comment]
Thinking out loud here, to enable the use of these elements without a warning, I could image a few contracts...

```tsx
function Scene() {
  const meshRef = useRef<THREE.Mesh>();

  useFrame(() => {
    if (!meshRef.current) return;
    meshRef.current.rotation.x += 0.01;
    meshRef.current.rotation.y += 0.01;
  });

  return (
    <>
      <ambientLight />
      <pointLight position={[10, 10, 10]} />
      <PerspectiveCamera makeDefault fov={75} position={[0, 0, 5]} />
      <mesh ref={meshRef}>
        <boxGeometry />
        <meshStandardMaterial color="hotpink" />
      </mesh>
    </>
  );
}
```

(Not final proposal, but basically this allows saying "pointLight can have position and the color property can be used for these"...)

```json
{
  "react/no-unknown-property": [
    "error",
    {
      "ignore": ["css"],
      "tags": {
        "pointLight": ["position"]
      },
      "properties": {
        "color": ["meshStandardMaterial"]
      }
    }
  ]
}
```

[Example for skia...](https://github.com/amilajack/react-skia-fiber)

```tsx
import React, { useEffect, useState } from "react";
import { SkParagraph } from "react-skia-fiber";

export default function App() {
  const paragraphRef = React.useRef<SkParagraph>();
  const [color, setColor] = useState("red");

  useEffect(() => {
    let colors = ["red", "green", "blue"];
    let i = 0;
    let interval = setInterval(() => {
      i++;
      setColor(colors[i % 3]);
    }, 1_000);
    return () => clearInterval(interval);
  }, []);

  return (
    <skCanvas clear="#ABACAB">
      <skParagraph
        x={100}
        y={100}
        width={500}
        ref={paragraphRef}
        text={color}
      />
      <skText x={100} y={100} text={color} />
    </skCanvas>
  );
}
```
Config...

```json
{
  "react/no-unknown-property": [
    "error",
    {
      "ignore": ["css"],
      "tags": {
        "skCanvas": ["clear"]
      },
      "properties": {
        "text": ["skText"]
      }
    }
  ]
}
```

Now this is pretty tedious for end users, but it would library authors to make their [plugins](https://github.com/pmndrs/react-three-fiber/tree/master/packages/eslint-plugin) better.

**Comment by brianzinn** [comment]
Thank-you @cruhl I like the constructive nature of this proposal.  For me the backwards compatible original behavior of ignoring unknown elements is best.

> Now this is pretty tedious for end users, but it would library authors to make their [plugins](https://github.com/pmndrs/react-three-fiber/tree/master/packages/eslint-plugin) better.

From my perspective having element names that can be ignored is enough.  Most libraries (likely?) have added their own global JSX IntrinsicElements, which are also typesafe and caught by modern IDEs.  ie: here is a light with optional numeric intensity in `react-babylonjs`:
![image](https://github.com/jsx-eslint/eslint-plugin-react/assets/542756/38d8460f-f4af-4b7b-ac17-fa6da7612f5d)

That's in addition to not being able to express typescript union and intersection types in a JSON format, etc.

Hope that helps - that's just one person's opinion.  I do appreciate your efforts.

**Comment by ljharb** [comment]
@cruhl let's consolidate it a bit - what about something where it's a single rule config keyed object, where the keys of the object are tag names, and the values are also an object, which has a "tag" string and a "properties" object (that lists the valid property names for that tag)?

it'd be a bit annoying if you have a single property you want added to a dozen tags, but i don't think config convenience is what's important here.

**Comment by karlhorky** [comment]
@cruhl @ljharb since this discussion is happening in an old, closed issue, should a new issue be opened with this proposal?

Or alternatively, should this issue be reopened?

**Comment by ljharb** [comment]
@karlhorky this one can be reopened once a suggested schema is confirmed to solve the issues discussed.
